UndertaleModLib/Scripting/IScriptInterface.cs

```diff
@@ -22,84 +24,469 @@ public ScriptException(string msg) : base(msg)
+        /// <param name="status">What the current status is. For example <c>Decompiling...</c>.</param>
+        /// <param name="progressValue">The value to set the progress bar to.</param>
+        /// <param name="maxValue">The max value of the progress bar.</param>
         void SetProgressBar(string message, string status, double progressValue, double maxValue);
+
+        /// <summary>
+        /// Show the progress bar.
+        /// </summary>
         void SetProgressBar();
+
+        /// <summary>
+        /// Updates the value of the current running progress bar dialog.
+        /// </summary>
+        /// <param name="progressValue">The new value to set the progress bar to.</param>
         void UpdateProgressValue(double progressValue);
+
+        /// <summary>
+        /// Updates the status of the current running progress bar dialog.
+        /// </summary>
+        /// <param name="status">The new status. For example <c>Decompiling...</c>.</param>
         void UpdateProgressStatus(string status);
+
+        //TODO: considering this forces everything that implements this to have their own progressValue,
+        //why not make that a necessary attribute?
+        /// <summary>
+        /// Adds a certain amount to the variable holding a progress value.
+        /// </summary>
+        /// <param name="amount">The amount to add.</param>
         void AddProgress(int amount);
-        void IncProgress();
-        void AddProgressP(int amount);
-        void IncProgressP();
+
+        /// <summary>
+        /// Increments the variable holding a progress value by one.
+        /// </summary>
+        void IncrementProgress();
+
+        /// <summary>
+        /// Adds a certain amount to the variable holding a progress value in.
+        /// Used for parallel operations, as it is thread-safe.
+        /// </summary>
+        /// <param name="amount">The amount to add.</param>
+        void AddProgressParallel(int amount);
+
+        /// <summary>
+        /// Increments the variable holding a progress value by one.
+        /// Used for parallel operations, as it is thread-safe.
+        /// </summary>
+        void IncrementProgressParallel();
+
+        /// <summary>
+        /// Gets the value of the variable holding a progress value.
+        /// </summary>
+        /// <returns>The value as <see cref="int"/>.</returns>
         int GetProgress();
+
+        /// <summary>
+        /// Sets the value of the variable holding a progress variable to another value.
+        /// </summary>
+        /// <param name="value">The new value for the progress variable.</param>
         void SetProgress(int value);
+
+        /// <summary>
+        /// Hides the progress bar.
+        /// </summary>
         void HideProgressBar();
+
+        /// <summary>
+        /// Enables the UI.
+        /// </summary>
         void EnableUI();
+
+        /// <summary>
+        /// Allows scripts to synchronize their assets with the lists from the UI.
```

Comment:
_"Allows scripts to modify asset lists from non-UI thread _(e.g. from within `Task.Run()`)_."_ would be better, because it describes better what it actually used for.

Reply:
Or maybe _"Calls `BindingOperations.EnableCollectionSynchronization()` for every listed asset list.
It allows scripts to modify asset lists from non-UI thread _(e.g. from within `Task.Run()`)_."_

Reply:
Disagree with the second comment, see description below as to why.

Regarding the first one, are they really modifying the asset list in the non-UI thread tho? As far as I can see they just modify their own assets which then get synced to the UI list.
In ImportGML.csx for example:
```cs
SyncBinding("Strings, Code, CodeLocals, Scripts, GlobalInitScripts, GameObjects, Functions, Variables", true);
await Task.Run(() => {
    foreach (string file in dirFiles)
    {
        IncrementProgress();

        ImportGMLFile(file, doParse, false, stopOnError);
    }
});
DisableAllSyncBindings();
```
Nothing actually accesses the list from the UI. `ImportGMLFile` changes assets inside of `Data` which then get synced with the UI list later because `SyncBinding` was called before.

Reply:
Most of asset lists in `Data` _(code, sprites, etc.)_ always bound to the _"Data"_ tree list.
And `ImportGMLFile()` modifies some of them while being within `Task.Run()` _(non-UI thread)_.
So, without `SyncBinding()`, it throws an error.

Reply:
Ah was not aware it throws an error otherwise.
Is 
`Allows scripts to modify asset lists from the non-UI thread. If this isn't called before attempting to modify them, an exception will be thrown.` 
fine?

Reply:
Yes, but _"... `System.NotSupportedException` will be thrown."_